rebasebot/bot.py

```diff
@@ -170,6 +172,70 @@ def _in_excluded_commits(sha: str, exclude_commits: list) -> bool:
     return False
 
 
+def _find_last_rebase_merge_commit(gitwd: git.Repo, source_repo: Repository, ancestry_path_merges) -> Commit:
+    logging.info("Searching for merge commit from previous rebasebot run to identify downstream commits")
+    pygit2_repo = pygit2.Repository(gitwd.working_tree_dir)
+    for merge_line in ancestry_path_merges:
+        sha, commit_message, committer_email = merge_line.split(" || ", 2)
+        logging.info(f"Checking: \"{commit_message}\"")
+        if "openshift-merge-bot[bot]@users.noreply.github.com" in committer_email:
+            logging.info("Skipping this merge as this bot only merges downstream")
+            # This skip is not required and is here only to improve performance.
+            continue
+        merge = gitwd.commit(sha)
+
+        # Now we get both parents of the merge commit and check if one of them is an ancestor of any
+        # of the upstream branches. If it is, we know that this merge commit is the last rebase merge commit
+        for parent in merge.parents:
+            for upstream_branch in source_repo.branches():
+                # When merge base is the same as parent, it means that the parent is an ancestor of the upstream branch
+                merge_base = pygit2_repo.merge_base(parent.hexsha, upstream_branch.commit.sha)
```

Comment:
I gather the `git` lib we are importing not able to perform this?
Is there any way in which we could achieve this without importing another lib?

Reply:
GitPython can do this, but it spawns the git process to get the result. pygit2 uses the libgit system library instead. This can be slow depending on how many merges and upstream branches the repository has. It was around 6 times slower. 5s vs 30s on kubernetes/autoscaler. Not a big deal in CI, but sometimes you want to run the rebasebot interactively.

Reply:
From the opposite side, how much work would it be to drop GitPython and move to pygit2? I assume more work than just this PR, but I'm not familiar enough with this code base to know how well the GitPython dependencies are wrapped between functions.

I ask because if we can get significant time savings for all git operations, that would be nice, and possibly worth switching to save on CI costs/developer time.

I share the concern of importing two different dependencies for roughly the same functionality, so standardizing on one would be great.

Reply:
Doing some more reading, pygit2 is an implementation of git plumbing, whereas GitPython wrap the git command for both plumbing and porcelain.

It's probably not worth re-implementing the high level stuff like checkouts ourselves within pygit2, not unless we can see even bigger than 6x speed ups.

So I am torn on including a new dependency here. It definitely saves a lot of time for interacting with bigger repos. It introduces the possibility that we'll be using different libraries inconsistently. If we keep it, I think we should maybe document the scenarios where using pygit2 is acceptable and default to GitPython most of the time.

Reply:
I was able to remove the library. 

Instead of iterating all branches and using git merge-base to figure out if commit is on that branch, we can use git branch --contains <commit>  to list all branches with that commit and then cheaply check if any upstream branch is in the list. 

This reduces git invocations from `len(merge_commits)*2(parents)*len(upstream_branches)` to `len(merge_commits)*2(parents)`.

This is sufficiently fast. In my testing 0.89s with the `if "openshift-merge-bot[bot]@users.noreply.github.com" in committer_email` skip and 4.7s without.

Reply:
Nice one @RadekManak 
Sounds like a very reasonable alternative

Reply:
Awesome, thanks for doing that research!